### ISSUE TITLE
issue with epoll_wait 

### ISSUE THREAD

**Issue by ravikumarbhattiprolu** [gramineproject/gramine#489]
## Description of the problem
epoll_wait times out without detecting the events. This was detected while using nginx in reverse proxy mode. The nginx forwards the requests to upstream server but does not recognize the response from upstream server. Captured the packets and saw that upstream server is receiving the request and responding. After debugging into the nginx code figured that epoll_wait is returning zero events and timing out. So, put together a sample test based on opensource socket programming examples to reproduce the epoll_wait issue.

## Steps to reproduce

1. Use the attached testcase epoll_issue.tar.gz
2. Extract the files into CI-Example directory
3. Build using make.
4. run the server normally. The server listens on port 50051
    %./server
5. run the client under gramine
    %gramine-direct ./helloworld
## Expected results
1. The client connects to the server, sends a hello and waits on epoll_wait.
2. server responds back with "hello from server". 
3. epoll_wait in the client recognizes the event and wakes up.
4. client calls recv to receive the message and print

## Actual results
In case of gramine execution, the epoll_wait times out and receives no events. 

## Gramine commit hash
It works only till commit c4c88f9bc1db53460774cd8adb07267236cef039 on 27th Jan 2021. We tried with latest from mainline, v1.1 and they did not work. 

[epoll_issue.tar.gz](https://github.com/gramineproject/gramine/files/8409232/epoll_issue.tar.gz)

**Comment by svenkata9** [comment]
@boryspoplawski Could you please take a look? This came from an internal team and blocking their release.

**Comment by boryspoplawski** [comment]
Ah, this is a known issue, same as #469 - it's actually issue with sockets.
I'm in the middle of rewriting whole sockets subsystem, but it will take some time. What release it's blocking? It's already on Gramine master

**Comment by boryspoplawski** [comment]
Ah, I misread `their` as `the`. Not sure what to do, it's unclear how to quickly solve this issue (and the rewrite will still take some time).

**Comment by ravikumarbhattiprolu** [comment]
The issue #469 was with accept call with sockets. This is with epoll handling. 
It is blocking one of our software releases where NGINX is used in reverse proxy mode with grpc server as upstream. With earlier Gramine v1.0 it is working fine. Now NGINX does not react to response from upstream server. We can't go back to v1.0 as there are other fixes happened after even v1.1 which we need.

**Comment by boryspoplawski** [comment]
> The issue https://github.com/gramineproject/gramine/issues/469 was with accept call with sockets. This is with epoll handling.

This is the same issue (the same bug is causing such behavior).
As a workaround you can change the order of functions call - in your example if you put `epoll_ctl(ADD)` **after** `connect` syscall, everything should work.

**Comment by ravikumarbhattiprolu** [comment]
We are installing NGINX in GSC container using apt install. We do not have the control on sources. Also figuring out the place to do this in NGINX code will not be easy. Is there any other option?

**Comment by boryspoplawski** [comment]
We have a fix for this, but it has potential to break other code. We are still discussing internally how to proceed with this

**Comment by aep** [comment]
for nginx, just disable epoll

```
events {
    use select;
}
```

**Comment by ravikumarbhattiprolu** [comment]
We are using nginx from Ubuntu rep using "apt install nginx". The stock ubuntu version of ngnix is not built with select support. "nginx -V' does not show --with-select_module configuration option. If we need to use this work around of using select, we need to build nginx from sources. Will try this. Need to change a few docker files.

**Comment by svenkata9** [comment]
@ravikumarbhattiprolu That's right, Ravi, but still the config file, like @aep mentioned can be updated before nginx runs - since this is related to nginx.conf

**Comment by ravikumarbhattiprolu** [comment]
@svenkata9 It is not just config file. When the nginx is built, it has to be configured (in ./configure script) with --with-select_module  parameter. Without that, if we add "use select;" in nginx.conf, the nginx throws an error and exits.

**Comment by ravikumarbhattiprolu** [comment]
@aep I tried both nginx-full and nginx-extras. Also, I tried installing from the nginx official packages (it installs latest 1.21.6). Neither of them have been built with select module. 
https://docs.nginx.com/nginx/admin-guide/installing-nginx/installing-nginx-open-source/ 
says: "The modules is built automatically if the platform does not appear to support more appropriate methods such as kqueue, epoll, or /dev/poll." Looks like epoll is more appropriate method...

Looks like building from the sources is only option
